streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java

```diff
@@ -170,7 +171,7 @@ private void putAndMaybeForward(final ThreadCache.DirtyEntry entry, final Intern
             final Bytes rawKey = Bytes.wrap(serdes.rawKey(key.key()));
             if (flushListener != null) {
                 final AGG newValue = serdes.valueFrom(entry.newValue());
-                final AGG oldValue = fetchPrevious(rawKey, key.window());
+                final AGG oldValue = sendOldValues ? fetchPrevious(rawKey, key.window()) : null;
```

Comment:
Here we need to do:
`final Agg oldValue == newValue == null || sendOldValues ? fetchPervious(..) : null;`
This is because `SessionWindows` have a dynamic time range, the the start is always fixed. So we need to send deletes for the previous smaller window when a window is merged, i.e, a simple count:
a@0  -> SessionKey(key=a start=0, end=0),  1
a@5  -> SessionKey(key=a start=0, end=0), null  (delete this as it is merged) 
               SessionKey(key=a start=0, end=5), 2   (this is the new merged session)